### PR TITLE
feat: add options validation

### LINKED ISSUE #9: Add options validation
Add validation of build federation service options

### PR DESCRIPTION
**Pull request by ianlnf**
- Adds options validation
- Updates readme

resolves #9

### CODE REVIEW

File: lib/federate.js [hunk 1]
```diff
@@ -249,13 +252,15 @@ function buildFederatedResolvers({ url, schema, resolvers }) {
...
-  // TODO validate args, handle errors
-
-  if (!options) {
-    options = { auto: true }
+async function buildFederatedService({ url, options: userOptions }) {
+  if (!url) {
+    throw new Error('buildFederatedService: url is required')
   }
 
+  const defaultOptions = { auto: true }
```
Comment (simone-sanfratello) [inline lib/federate.js:260]: `defaultOptions` can be defined outside of the function

File: lib/federate.js [hunk 2]
```diff
@@ -249,13 +252,15 @@ function buildFederatedResolvers({ url, schema, resolvers }) {
   return federatedResolvers
 }
 
-async function buildFederatedService({ url, options }) {
-  // TODO validate args, handle errors
-
-  if (!options) {
-    options = { auto: true }
+async function buildFederatedService({ url, options: userOptions }) {
+  if (!url) {
```
Comment (simone-sanfratello) [inline lib/federate.js:256]: I'd move this in a validation function too, and check also for type string

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/validators.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      )
+    }
+
+    Object.keys(options.resolvers).forEach(resolver => {
+      validateResolverObject(resolver, options.resolvers[resolver])
+    })
+  }
+}
+
+function validateUrl(url) {
```
Comment (simone-sanfratello) [inline lib/validators.js:75]: validate the url with `URL` class

File: package.json [hunk 4]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "lint": "eslint .",
     "test": "tap test/*.test.js",
+    "test:watch": "tap test/*.test.js -w",
```
[the remark this answers is not in the document]
Comment (simone-sanfratello) [reply package.json:9]: please remove this

File: lib/validators.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+// Validates a single resolver object, e.g. { __resolveReference: () => {} }
+function validateResolverObject(resolverKey, resolverObject) {
+  if (typeof resolverObject !== 'object') {
+    throw new Error(
+      `buildFederatedService: options resolvers '${resolverKey}' value must be an object`
+    )
+  }
+
+  if (resolverObject.__resolveReference == null) {
```
Comment (simone-sanfratello) [inline lib/validators.js:40]: `__resolveReference` is not mandatory

File: test/options.test.js [hunk 6]
```diff
@@ -1,232 +1,84 @@
...
+  },
+  {
+    name: 'resolvers entry must be an object',
+    options: { resolvers: { User: '' } },
+    expected: `options resolvers 'User' value must be an object`
+  },
+  {
+    name: 'resolvers entry must have a reference resolver',
+    options: { resolvers: { User: {} } },
+    expected: `options resolvers 'User' value must have a reference resolver`
```
Comment (simone-sanfratello) [inline test/options.test.js:61]: add tests for the url, since there is a `validateUrl` function

**Review by simone-sanfratello — CHANGES_REQUESTED** [review]

File: lib/validators.js (hunk 5 above)
Comment (krisgrint) [reply lib/validators.js]: I have removed this check

File: test/options.test.js [hunk 7]
```diff
@@ -1,232 +1,84 @@
 'use strict'
 
 const t = require('tap')
-const dedent = require('dedent')
-const graphql = require('graphql')
+const mergeOptions = require('merge-options').bind({ ignoreUndefined: true })
 
-const { autoOptions } = require('../lib/options')
```
[the remark this answers is not in the document]
Comment (krisgrint) [reply test/options.test.js:7]: they have been moved to `test/auto-options.test.js`

**Review by simone-sanfratello — APPROVED** [review]

**Pull request merged**